File: Android/service/src/main/java/com/scurab/android/anuitor/nanoplugin/AggregateMimePlugin.java
package com.scurab.android.anuitor.nanoplugin;

import java.io.File;
import java.util.HashSet;
import java.util.Map;

import fi.iki.elonen.NanoHTTPD;

/**
 * Created by jbruchanov on 20/05/2014.
 * Simple wrapper for multiple plugins with same mime type.
 */
public class AggregateMimePlugin extends BasePlugin {

    private BasePlugin[] mPlugins;
    private String[] mFiles;

    public AggregateMimePlugin(BasePlugin... plugins) {
        mPlugins = plugins;
        if (mPlugins == null || mPlugins.length == 0) {
            throw new IllegalArgumentException("Plugins are null or empty!");
        }
        checkMimeTypes();
        mFiles = mergeFiles(mPlugins);
    }

    private String[] mergeFiles(BasePlugin[] plugins) {
        HashSet<String> files = new HashSet<String>();
        for (BasePlugin b : plugins) {
            for (String file : b.files()) {
                if(files.contains(file)){
                    throw new IllegalStateException(String.format("File:'%s' is already defined from previous plugin", file));
                }
                files.add(file);
            }
        }
        return files.toArray(new String[files.size()]);
    }

    private void checkMimeTypes() {
        String mime = mPlugins[0].mimeType();
        for (BasePlugin b : mPlugins) {
            if (!mime.equals(b.mimeType())) {
                throw new IllegalStateException(String.format("Plugins don't have same mime types %s vs %s", mime, b.mimeType()));
            }
        }
    }

    @Override
    public String[] files() {
        return mFiles;
    }

    @Override
    public String mimeType() {
        return mPlugins[0].mimeType();
    }

    @Override
    public boolean canServeUri(String uri, File rootDir) {
        return getServeCandidate(uri, rootDir) != null;
    }

    @Override
    public NanoHTTPD.Response serveFile(String uri, Map<String, String> headers, NanoHTTPD.IHTTPSession session, File file, String mimeType) {
        //null should never happened, because it's not called if canServerUri returns false
        return getServeCandidate(uri, file /*TODO: check if it's fine */).serveFile(uri, headers, session, file, mimeType);
    }

    BasePlugin getServeCandidate(String uri, File rootDir) {
        for (BasePlugin plugin : mPlugins) {
            if (plugin.canServeUri(uri, rootDir)) {
                return plugin;
            }
        }
        return null;
    }
}

File: Android/service/src/main/java/com/scurab/android/anuitor/nanoplugin/AggregateMimePlugin.kt
package com.scurab.android.anuitor.nanoplugin

import fi.iki.elonen.NanoHTTPD
import java.io.File
import java.lang.NullPointerException
import java.util.HashSet

class AggregateMimePlugin(vararg plugins: BasePlugin) : BasePlugin() {

    private val files: Array<String>
    private val plugins: Array<out BasePlugin>
    private val mimeType: String

    override fun files(): Array<String> = files
    override fun mimeType(): String = mimeType

    init {
        if (plugins.isEmpty()) {
            throw IllegalArgumentException("Plugins can't be empty!")
        }
        this.plugins = plugins
        checkMimeTypes()
        files = mergeFiles(plugins)
        mimeType = plugins.first().mimeType()
    }

    override fun canServeUri(uri: String, rootDir: File): Boolean {
        return getServeCandidate(uri, rootDir) != null
    }

    override fun serveFile(uri: String, headers: Map<String, String>, session: NanoHTTPD.IHTTPSession, file: File, mimeType: String): NanoHTTPD.Response {
        //null should never happened, because it's not called if canServerUri returns false
        return (getServeCandidate(uri, file /*TODO: check if it's fine */)
                ?: throw NullPointerException("Unable to handle uri:'$uri'"))
                .serveFile(uri, headers, session, file, mimeType)
    }

    fun getServeCandidate(uri: String, rootDir: File): BasePlugin? {
        return plugins.firstOrNull { it.canServeUri(uri, rootDir) }
    }

    private fun mergeFiles(plugins: Array<out BasePlugin>): Array<String> {
        val files = HashSet<String>()
        for (b in plugins) {
            for (file in b.files()) {
                if (files.contains(file)) {
                    throw IllegalStateException("File:'$file' is already defined from previous plugin")
                }
                files.add(file)
            }
        }
        return files.toTypedArray()
    }

    private fun checkMimeTypes() {
        plugins.map { it.mimeType() }
                .toSet()
                .let {
                    if (it.size != 1) {
                        throw IllegalStateException("Plugins have to have same mime type. MimeType:$it")
                    }
                }

    }
}